// nextra-theme-blog/dist/cusdis.js
import {
  useBlogContext
} from "./chunk-AE2OX3A7.js";
import "./chunk-PKUHTIDK.js";

// src/cusdis.tsx
import { useTheme } from "next-themes";
import dynamic from "next/dynamic";
import { useRouter } from "next/router";
import React from "react";
var Cusdis = dynamic(
  () => import("react-cusdis").then((mod) => mod.ReactCusdis),
  { ssr: false }
);
var Comments = () => {
  const { config, opts } = useBlogContext();
  const router = useRouter();
  const { resolvedTheme } = useTheme();
  const { cusdis } = config;
  if (!cusdis) {
    return null;
  }
  if (!cusdis.appId) {
    console.warn("[nextra/cusdis] `appId` is required");
    return null;
  }
  return /* @__PURE__ */ React.createElement(Cusdis, {
    lang: cusdis.lang,
    style: { marginTop: "4rem" },
    attrs: {
      host: cusdis.host || "https://cusdis.com",
      appId: cusdis.appId,
      pageId: router.pathname,
      pageTitle: opts.title,
      theme: resolvedTheme === "dark" ? "dark" : "light"
    }
  });
};
var cusdis_default = Comments;
export {
  cusdis_default as default
};
